### DISCUSSION TITLE
Manipulating (pre-)existing thin-pool devices not backed by `LinearDev`s

### DISCUSSION THREAD

**Discussion by ckatsak** [stratis-storage/devicemapper-rs#945] [Q&A] [answered]
Hello! Thank you for maintaining this work :)

As far as I can tell, the high-level crate (`devicemapper`) cannot be employed to manipulate (pre-)existing thin-pool devices present in the system, unless they are backed by `LinearDev` block devices. Otherwise, no corresponding `ThinPoolDev` can be instantiated (e.g., via `ThinPoolDev::setup()`).

For example, no valid `ThinPoolDev` can be instantiated for the following:
```console
# dmsetup table dm-thinpool -v
Name:              dm-thinpool
State:             ACTIVE
Read Ahead:        256
Tables present:    LIVE & INACTIVE
Open count:        0
Event number:      0
Major, minor:      254, 0
Number of targets: 1

0 104857600 thin-pool 7:1 7:0 128 32768 1 skip_block_zeroing
```
since its data & metadata devices are loop devices (major number 7).

Is that right, or am I missing something?

I hope I'm asking this at the right place. Thanks!

**Comment by mulkieran** [comment, marked as answer]
@ckatsak That is correct. And it is clear that it is not convenient. However, you might still be able to use the library in one of two ways in order to do this. First, this project exports lower-level -sys bindings, which you can use independently: https://crates.io/crates/devicemapper-sys . Second, this library supports a core module which is built on the lower-level bindings but is not DM target-specific. While it removes thin pool specific conveniences that you might prefer to have available, it might allow you more flexibility to do what it is that you are trying.

**Reply by ckatsak** [reply]
Great, thanks!
